src/ParallelAlgorithms/NonlinearSolver/NewtonRaphson/ResidualMonitor.hpp

```diff
@@ -0,0 +1,106 @@
+// Distributed under the MIT License.
+// See LICENSE.txt for details.
+
+#pragma once
+
+#include <limits>
+
+#include "AlgorithmSingleton.hpp"
+#include "DataStructures/DataBox/DataBox.hpp"
+#include "DataStructures/DataBox/Prefixes.hpp"
+#include "IO/Observer/Actions/RegisterSingleton.hpp"
+#include "Informer/Tags.hpp"
+#include "NumericalAlgorithms/Convergence/Tags.hpp"
+#include "Parallel/GlobalCache.hpp"
+#include "ParallelAlgorithms/Initialization/MergeIntoDataBox.hpp"
+#include "ParallelAlgorithms/LinearSolver/Tags.hpp"
+#include "ParallelAlgorithms/NonlinearSolver/Observe.hpp"
+#include "ParallelAlgorithms/NonlinearSolver/Tags.hpp"
+
+/// \cond
+namespace tuples {
+template <typename...>
+class TaggedTuple;
+}  // namespace tuples
+namespace NonlinearSolver::newton_raphson::detail {
+template <typename Metavariables, typename FieldsTag>
+struct InitializeResidualMonitor;
+}  // namespace NonlinearSolver::newton_raphson::detail
+/// \endcond
+
+namespace NonlinearSolver::newton_raphson::detail {
+
+template <typename Metavariables, typename FieldsTag, typename OptionsGroup>
+struct ResidualMonitor {
+  using chare_type = Parallel::Algorithms::Singleton;
+  using const_global_cache_tags =
+      tmpl::list<logging::Tags::Verbosity<OptionsGroup>,
+                 Convergence::Tags::Criteria<OptionsGroup>,
+                 NonlinearSolver::Tags::SufficientDecrease<OptionsGroup>,
+                 NonlinearSolver::Tags::MaxGlobalizationSteps<OptionsGroup>>;
+  using metavariables = Metavariables;
+  using phase_dependent_action_list = tmpl::list<
+      Parallel::PhaseActions<
+          typename Metavariables::Phase, Metavariables::Phase::Initialization,
+          tmpl::list<InitializeResidualMonitor<FieldsTag, OptionsGroup>>>,
+      Parallel::PhaseActions<
+          typename Metavariables::Phase,
+          Metavariables::Phase::RegisterWithObserver,
+          tmpl::list<observers::Actions::RegisterSingletonWithObserverWriter<
+              NonlinearSolver::observe_detail::Registration<OptionsGroup>>>>>;
+
+  using initialization_tags = Parallel::get_initialization_tags<
+      Parallel::get_initialization_actions_list<phase_dependent_action_list>>;
+
+  static void initialize(
+      Parallel::CProxy_GlobalCache<Metavariables>& /*global_cache*/) noexcept {}
+
+  static void execute_next_phase(
+      const typename Metavariables::Phase next_phase,
+      Parallel::CProxy_GlobalCache<Metavariables>& global_cache) noexcept {
+    auto& local_cache = *(global_cache.ckLocalBranch());
+    Parallel::get_parallel_component<ResidualMonitor>(local_cache)
+        .start_phase(next_phase);
+  }
+};
+
+template <typename FieldsTag, typename OptionsGroup>
+struct InitializeResidualMonitor {
```

Comment:
- Add documentation
- for my understanding, why is this not in the ResidualMonitorActions file?

Reply:
- See https://github.com/sxs-collaboration/spectre/pull/2564#discussion_r526164379 for why don't think this action needs a Doxygen entry. Regarding maintainability: I think this action is self-explanatory and doesn't need a code comment. Please let me know if you disagree.
- No reason other than consistency: I started placing the init-databox actions in the `ResidualMonitor.hpp` file for the linear solvers for some reason, so did the same here. They could all move to their respective `ResidualMonitorActions.hpp` file. I don't feel strongly either way 🤷‍♂️

Reply:
👍 I'm satisfied as-is